### DISCUSSION TITLE
Wire.h ... setup for Master and Slave on DA28 (TWI_MANDS_SINGLE), etc?

### DISCUSSION THREAD

**Discussion by gordonthree** [SpenceKonde/DxCore#312] [Library discussion]
Sorry, platformio user here, so I miss out on some of the gui selection goodness of the official IDE.

I am developing on an AVR128DA32 and AVR128DA28 chips, using them as i2c slaves in battery monitoring boards. I have a very nice 24bit 2 channel ADC, that is also i2c, and only has a single fixed address. My goal is having multiple DA chips talking i2c to a master, probably ESP32, which is recording everything to sd card. The DA chips are also doing other monitoring, providing basic BMS type functions and coulomb counting for state of charge.

So, I would like to figure out Master and Slave mode on the DA28, using different pins. Here's what I have so far:

```
  #if defined(TWI_MORS_BOTH)
    Wire1.begin(); // master mode on TWI1 pins, no alternatives available on the 32 pin chip
    Wire1.setClock(100000);  // bus speed 100khz

    //Wire.pins(SDA0, SCL0); // slave on preferred TWI0 pins, compiler throws an error that I'm not allowed to change pins
    Wire.begin(I2C_SLAVE_ADDR); 
  #else 
    // No TWI1, so setup TWI0 for dual mode ... TWI_MANDS_SINGLE
    // Wire1.pins(SDA1, SCL1);
    Wire.begin(SDA1,SCL1); // master on secondary pins (twi1 for the 32 pin chip)
    Wire.setClock(100000);

    Wire.pins(SDA0, SCL0); // slave on primary pins
    Wire.begin(I2C_SLAVE_ADDR); // slave on twi0
  #endif
```

First off, I'm not sure that Wire1 means TWI1, or what? When I try to set pins on Wire1, it complains that pin change is not permitted on 32pin chips.

If I try to use Wire.pins to change pins it also complains. So maybe my assumption of what Wire and Wire1 reference are incorrect.

The code as posted (with the .pins) functions commented out compiles, as long as I add build flag TWI_MORS_BOTH for the 32pin and TWI_MANDS_SINGLE for the 28 pin.

I don't have either chip yet, hopefully this weekend. Any suggestions are welcome. 

Full code and project [here](https://github.com/gordonthree/packmonitor/tree/AVR-DA)

**Comment by SpenceKonde** [comment]
Wire1 is TWI1. 

You cannot change the pins on it however, please refer to section 3.6 of the PORTMUX chapter. 
The options are 
* PF2, PF3 + PB2, PB3
* PF2, PF3 + PB6, PB7
* PB2, PB3 + PB6, PB7
Since there is no port B on 32-pin chips (much less PB6 and 7 which we don't see until 64 pin parts), the only option for TWI1 is non-dual mode using pins F2 and F3. 


If nowhere in the program are the pins changed, there is no reason ro change it from it's defauLt value anyway. 

Umm to you have warnings enabled? If so either I've been less successful at making them impossible to suppress than I thought, or you should run with warnings enabled. 

(also per docs, swap() is preferable to pins().

You rely on SDA0 and SCL0. This core doesn't have those, it has SCL and SDA for those pins. What's the standard behavior here? I thought they only added the number after the first one - but if otherboards have SCL0 and SDA0 then it's a bug that I don't define them.

**Reply by gordonthree** [reply]
That makes perfect sense for TWI1.

If I am running TWI_MANDS_SINGLE on the 28pin, what would the config steps look like to setup pins PA2/PA3 for the master role and PC2/3 for the slave role, rather than have both roles on the same pins?

I appreciate the help, thank you!

**Comment by gordonthree** [comment]
The compiler seems happy with this, for both the 28pin and 32pin with their respective flags set in platformio.ini. Will find out if it works next week.

Does this look proper?

```
#if defined(TWI_MORS_BOTH)
TwoWire i2c_slave(&TWI0);
TwoWire i2c_master(&TWI1);
#elif defined(TWI_MANDS_SINGLE)
TwoWire i2c_slave(&TWI0);
TwoWire i2c_master(&TWI0);
#endif

void setup() {
  #if defined(TWI_MORS_BOTH)
    // i2c_master is TWI1, using default pins PF2, PF3
    i2c_master.begin(); 
    i2c_master.setClock(100000);  // bus speed 100khz

    // i2c_slave is TWI0, accept default pins, should be PA2, PA3
    i2c_slave.begin(I2C_SLAVE_ADDR); 
  #else 
    // Setup TWI0 for dual mode ... TWI_MANDS_SINGLE
    i2c_master.begin(PIN_PC2, PIN_PC3); // master on alternate pins
    i2c_master.setClock(100000); // 100khz clock

    // Setup second instance of TWO0 as slave
    i2c_slave.pins(PIN_PA2, PIN_PA3); // slave on alternate pins
    i2c_slave.begin(I2C_SLAVE_ADDR); // slave on twi0
  #endif

  #ifdef TWI_MANDS_SINGLE 
    #pragma message "TWI_MANDS_SINGLE defined!"
  #endif

  delay(2000);
```
